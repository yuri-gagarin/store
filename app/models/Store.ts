import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;