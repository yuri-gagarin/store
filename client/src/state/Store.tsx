import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

