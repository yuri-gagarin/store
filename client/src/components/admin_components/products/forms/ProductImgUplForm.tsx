import React, { useState, useContext } from "react";
import { Button, Form } from "semantic-ui-react";
// css imports //
import "./css/productImgUploadForm.css";
// actions //
import { uploadProductImage } from "../actions/APIProductActions";
// state //
import { Store } from "../../../../state/Store";

const ProductImageUplForm: React.FC<{}> = (props): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const [ file, setFile ] = useState<File>();
  const { currentProductData } = state.productState;
  const handleButtonClick = () => {

  };
  const uploadFile = () => {
    if (file) {
      const formData = new FormData();
      const { _id } = currentProductData;
      formData.append("productImage", file);
      uploadProductImage(_id, formData, state, dispatch)
        .then((success) => {
          if (success) {
            setFile(undefined)
          }
        })
    }
  }
  const cancelFile = () => {
    setFile(undefined);
  }

  const fileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0])
    }
  };

  return (
    <div id="productImgUplFormHolder">
      <div><p>Image Uploader</p></div>
      <Form id="productImgUploadForm">
        {
          !file ?
          <div id="productImgInputControlls"> 
            <Button
              as="label"
              content="Choose Image"
              labelPosition="left"
              htmlFor="productImgFile"
              icon="file"
              onClick={handleButtonClick}
            />
            <input type="file" id="productImgFile" hidden onChange={fileChange} />
          </div>
          : null
        }
        
        {
          file ?  
          <div id="productImgUploadControlls">
            <Button 
              id="productImgCancelBtn"
              as="label"
              content="Cancel"
              labelPosition="left"
              icon="cancel"
              onClick={cancelFile}
            />
            <Button 
              id="productImgUploadBtn"
              as="label"
              content="Upload"
              icon="upload"
              onClick={uploadFile}
            />
          </div>
          : null
        }
       

       <span>{file ? file.name : "No File"}</span>
      </Form>
    </div>
  );
};

export default ProductImageUplForm;

