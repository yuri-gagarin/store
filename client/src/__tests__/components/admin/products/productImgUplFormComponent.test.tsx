import React from "react";
import { Button } from "semantic-ui-react";
import moxios from "moxios";
// testing imports //
import { mount, ReactWrapper } from "enzyme";
import { act } from 'react-dom/test-utils';
// components //
import ProductImgUplForm from "../../../../components/admin_components/products/forms/ProductImgUplForm";
// React.Context and State //
import { TestStateProvider } from "../../../../state/Store";
// helpers //
import MockFile from "../../../helpers/mockFile";
import { createMockProductImage, createMockProducts } from "../../../../test_helpers/productHelpers";
import ProductImageUplForm from "../../../../components/admin_components/products/forms/ProductImgUplForm";
import { generateCleanState } from "../../../../test_helpers/miscHelpers";

describe("Product Image Upload Form Tests", () => {
  let mockProductData: IProductData;
  
  beforeAll(() => {
    mockProductData = createMockProducts(1)[0];
    mockProductData.images[0] = createMockProductImage(mockProductData._id);
  });
  
  describe("Render tests without any Image data", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      wrapper = mount(<ProductImgUplForm />)
     });

    it("Should properly render", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render one '#productImgSelectBtn' element", () => {
      const uplButton = wrapper.render().find("#productImgSelectBtn");
      expect(uplButton.length).toEqual(1);
    });
    it("Should NOT render '#productImgUploadControls", () => {
      const controls = wrapper.find("#productImgUploadControls");
      expect(controls.length).toEqual(0);
    });
    it("Should render an input", () => {
      const input = wrapper.find("input");
      expect(input.length).toEqual(1);
    });
  });
  
  describe("Render tests with an Image file present", () => {
    let wrapper: ReactWrapper;
    let input: ReactWrapper;

    beforeAll(() => {
      wrapper =  mount(<ProductImgUplForm />);
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = wrapper.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    it("Should properly set the file in props", () => {
      expect(input.prop("type")).toBeDefined();
      expect(input.prop("type")).toEqual("file");
    });
    it("Should successfully update and render the '#productImgUploadBtn' button", () => {
      const imgUpoadBtn = wrapper.render().find("#productImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should successfully update and render the '#productImgCancelBtn' button", () => {
      const imgCanceldBtn = wrapper.render().find("#productImgCancelBtn");
      expect(imgCanceldBtn.length).toEqual(1);
    });
    it("Should NOT render the '#productImageRetryButton'", () => {
      const retryButton  = wrapper.render().find("#productImgRetryButton");
      expect(retryButton.length).toEqual(0);
    });
    it("Should properly handle the '#productImgCancelBtn' and render correct local state", () => {
      const imgCancelBtn = wrapper.find("#productImgCancelBtn");
      imgCancelBtn.at(0).simulate("click");
      expect(wrapper.render().find("#productImgCancelBtn").length).toEqual(0);
      expect(wrapper.render().find("#productImgUploadBtn").length).toEqual(0);
      expect(wrapper.render().find("#productImgRetryBtn").length).toEqual(0);
      expect(wrapper.render().find("#productImgSelectBtn").length).toEqual(1);
    });
  });
  
  // MOCK Successful Image upload tests //
  describe("'#productImgUploadBtn' functionality and successful upload with local state changes", () => {
    let component: ReactWrapper;
    let input: ReactWrapper;

    beforeAll(() => {
      const state = generateCleanState();
      state.productState.currentProductData = mockProductData;

      moxios.install();
      moxios.stubRequest(`/api/uploads/product_images/${mockProductData._id}`, {
        status: 200,
        response: {
          responseMsg: "ok",
          updatedProduct: mockProductData
        }
      });

      component = mount(
        <TestStateProvider mockState={state}>
          <ProductImgUplForm />
        </TestStateProvider>
      );
      // mock an image //
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = component.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    afterAll(() => {
      moxios.uninstall();
    });
    it("Should dispatch a 'correct' API request show 'loader'", async () => {
      const promise = Promise.resolve();
      const imgUpoadBtn = component.find("#productImgUploadBtn");
      imgUpoadBtn.at(0).simulate("click");
      // clear promises //
      await act( async () => promise);
      const uploadButton = component.find(ProductImageUplForm).find(Button).at(1);
      expect(uploadButton.props().loading).toEqual(true);
    });

    it("Should correctly render '#productImgSelectBtn' button after 'successful upload", () => {
      component.update();
      const selectImgBtn = component.find(ProductImgUplForm).render().find("#productImgSelectBtn");
      expect(selectImgBtn.length).toEqual(1);
    });
    it("Should NOT render '#productImgRetryButton' after 'successful upload", () => {
      const imgUpoadBtn = component.find(ProductImgUplForm).render().find("#productImgRetryButton");
      expect(imgUpoadBtn.length).toEqual(0);
    });
    it("Should NOT render '#productImgUploadBtn' button after 'successful upload", () => {
      const imgUpoadBtn = component.find(ProductImgUplForm).render().find("#productImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    });
    it("Should NOT render '#productImgCancelBtn' button after 'successful upload", () => {
      const imgUpoadBtn = component.find(ProductImgUplForm).render().find("#productImgCancelBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    });
  });
  // END Mock successful Image upload tests //
  // MOCK unsuccessful Image upload tests //
  
  describe("'#cancelProductImgUploadBtn' functionality and a failed upload", () => {
    let wrapper: ReactWrapper;
    let input: ReactWrapper;
    const error = new Error("An error occured");

    beforeAll(() => {
      const state = generateCleanState();
      state.productState.currentProductData = mockProductData;

      moxios.install();
      moxios.stubRequest(`/api/uploads/product_images/${mockProductData._id}`, {
        status: 500,
        response: {
          responseMsg: "Error",
          error: error
        }
      });

      wrapper = mount(
        <TestStateProvider mockState={state}>
          <ProductImgUplForm />
        </TestStateProvider>
      );
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = wrapper.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });

    afterEach(() => {
      moxios.uninstall();
    });

    it("Should successfully handle the '#productImgUploadBtn' API error, render 'loader'", async () => {
      const promise = Promise.resolve();
      const imgUpoadBtn = wrapper.find("#productImgUploadBtn").at(0);

      imgUpoadBtn.simulate("click");
      // loader should be displayed on the  #productImgUploadBtn //
      await act( async() => promise);
      const updatedButton = wrapper.find(ProductImageUplForm).find(Button).at(1);
      expect(updatedButton.props().loading).toEqual(true);
    });
    it("Should render '#productImgRetryButton' button after a 'failed' upload", () => {
      const imgUpoadBtn = wrapper.find(ProductImgUplForm).render().find("#productImgRetryButton");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should render '#productImgCancelBtn' button after a 'failed' upload", () => {
      const imgUpoadBtn = wrapper.find(ProductImgUplForm).render().find("#productImgCancelBtn");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should NOT render '#selectProductImgBtn' button after a 'failed' upload", () => {
      const selectImgBtn = wrapper.find(ProductImgUplForm).render().find("#selectProductImgBtn");
      expect(selectImgBtn.length).toEqual(0);
    });
    
    
  });
  
  // END Mock unsuccessful Image upload tests //
});