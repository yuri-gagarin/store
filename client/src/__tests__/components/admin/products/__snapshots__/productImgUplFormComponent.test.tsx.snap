// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Image Upload Form Tests Render tests without any Image data Should properly render 1`] = `
<ProductImageUplForm>
  <div
    id="productImgUplFormHolder"
  >
    <div>
      <p>
        Image Uploader
      </p>
    </div>
    <Form
      as="form"
      id="productImgUploadForm"
    >
      <form
        className="ui form"
        id="productImgUploadForm"
        onSubmit={[Function]}
      >
        <div
          id="productImgInputControlls"
        >
          <Button
            as="label"
            content="Choose Image"
            htmlFor="productImgFile"
            icon="file"
            id="selectProductImgBtn"
            labelPosition="left"
            onClick={[Function]}
          >
            <Ref
              innerRef={
                Object {
                  "current": <label
                    class="ui icon left labeled button"
                    for="productImgFile"
                    id="selectProductImgBtn"
                    role="button"
                  >
                    <i
                      aria-hidden="true"
                      class="file icon"
                    />
                    Choose Image
                  </label>,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <label
                      class="ui icon left labeled button"
                      for="productImgFile"
                      id="selectProductImgBtn"
                      role="button"
                    >
                      <i
                        aria-hidden="true"
                        class="file icon"
                      />
                      Choose Image
                    </label>,
                  }
                }
              >
                <label
                  className="ui icon left labeled button"
                  htmlFor="productImgFile"
                  id="selectProductImgBtn"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="file"
                  >
                    <i
                      aria-hidden="true"
                      className="file icon"
                      onClick={[Function]}
                    />
                  </Icon>
                  Choose Image
                </label>
              </RefFindNode>
            </Ref>
          </Button>
          <input
            hidden={true}
            id="productImgFile"
            onChange={[Function]}
            type="file"
          />
        </div>
        <span>
          No File
        </span>
      </form>
    </Form>
  </div>
</ProductImageUplForm>
`;
